File: frontend/src/engine/capabilities/insertDeleteColumns.ts
/**
 * insertDeleteColumns — parity with insertDeleteRows.
 *
 * Accepts a column-letter range like "C:E" (or a cell-address range whose
 * column span we infer) and either inserts blank columns (shifting existing
 * ones) or deletes columns (shifting neighbors).
 */

import { CapabilityMeta, InsertDeleteColumnsParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";
import { registerInverseOp } from "../snapshot";

const meta: CapabilityMeta = {
  action: "insertDeleteColumns",
  description: "Insert or delete worksheet columns — column-axis pair of insertDeleteRows",
  mutates: true,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.1",
};

async function handler(
  context: Excel.RequestContext,
  params: InsertDeleteColumnsParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { range: address, action, shiftDirection = "right" } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would ${action} columns ${address}` };
  }

  options.onProgress?.(`${action === "insert" ? "Inserting" : "Deleting"} columns ${address}...`);

  // If the user gave a cell-address range like "C1:E10", getEntireColumn() on
  // that range resolves to the columns C:E. For a plain "C:E" the range is
  // already full-column.
  const range = resolveRange(context, address);
  const colRange = range.getEntireColumn();
  colRange.load(["columnCount", "address"]);
  await context.sync();

  try {
    if (action === "insert") {
      const dir = shiftDirection === "left"
        ? Excel.InsertShiftDirection.right // "shift existing cells right" = inserting to the LEFT of them
        : Excel.InsertShiftDirection.right;
      // For whole-column inserts the direction is moot (you're inserting full columns);
      // we keep the param for API symmetry with insertDeleteRows.
      colRange.insert(dir);
      await context.sync();
      // Undo an insert = delete the columns we just added.
      colRange.load(["address", "worksheet/name"]);
      await context.sync();
      const sheetName = colRange.worksheet.name;
      const rangeAddr = colRange.address.includes("!") ? colRange.address.split("!").pop()! : colRange.address;
      registerInverseOp({ kind: "deleteColumns", sheetName, rangeAddress: rangeAddr });
    } else {
      colRange.delete(Excel.DeleteShiftDirection.left);
      await context.sync();
    }
  } catch (err: unknown) {
    const msg = err instanceof Error ? err.message : String(err);
    return {
      stepId: "",
      status: "error",
      message: `Failed to ${action} columns ${address}: ${msg}`,
      error: msg,
    };
  }

  return {
    stepId: "",
    status: "success",
    message: `${action === "insert" ? "Inserted" : "Deleted"} columns ${address}.`,
    outputs: { range: address, columnCount: colRange.columnCount },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/insert_delete_columns.py
"""insertDeleteColumns — insert or delete worksheet columns.

Parity with insert_delete_rows.py. Accepts a column-letter range ("C:E") or
a cell-address range and uses its column span via COM's `EntireColumn`.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    address = params.get("range")
    action = params.get("action")
    shift_direction = params.get("shiftDirection", "right")

    if not address:
        return {"status": "error", "message": "insertDeleteColumns requires 'range'."}
    if action not in ("insert", "delete"):
        return {"status": "error", "message": "'action' must be 'insert' or 'delete'."}

    rng = resolve_range(ctx.workbook_handle, address)

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would {action} columns {rng.address}.",
        }

    try:
        col_range = rng.api.EntireColumn
        if action == "insert":
            # xlShiftToRight = -4161, xlShiftDown = -4121 (irrelevant here).
            col_range.Insert()
        else:
            col_range.Delete()
    except Exception as exc:  # noqa: BLE001
        return {
            "status": "error",
            "message": f"{action.capitalize()} columns failed on {rng.address}: {exc}",
            "error": str(exc),
        }

    return {
        "status": "success",
        "message": f"{action.capitalize()}ed columns {rng.address} (shift {shift_direction}).",
        "outputs": {"range": rng.address, "columnCount": rng.columns.count},
    }


registry.register("insertDeleteColumns", handler, mutates=True)
